src/btf.cpp

```diff
@@ -38,151 +39,116 @@ static __u32 type_cnt(const struct btf *btf)
 #endif
 }
 
-static unsigned char *get_data(const char *file, ssize_t *sizep)
+__s32 BTF::start_id(const struct btf *btf) const
 {
-  struct stat st;
-
-  if (stat(file, &st))
-    return nullptr;
-
-  FILE *f;
-
-  f = fopen(file, "rb");
-  if (!f)
-    return nullptr;
+  return btf == vmlinux_btf ? 1 : vmlinux_btf_size + 1;
+}
 
-  unsigned char *data;
-  unsigned int size;
+static int libbpf_print(enum libbpf_print_level level,
+                        const char *msg,
+                        va_list ap)
+{
+  fprintf(stderr, "BTF: (%d) ", level);
+  return vfprintf(stderr, msg, ap);
+}
 
-  size = st.st_size;
+BTF::BTF(const std::set<std::string> &modules) : state(NODATA)
+{
+  if (bt_debug != DebugLevel::kNone)
+    libbpf_set_print(libbpf_print);
 
-  data = (unsigned char *) malloc(size);
-  if (!data)
+  // Try to get BTF file from BPFTRACE_BTF env
+  char *path = std::getenv("BPFTRACE_BTF");
+  if (path)
   {
-    fclose(f);
-    return nullptr;
+    btf_objects.emplace_back(btf__parse_raw(path), 0, "");
+    vmlinux_btf = btf_objects.back().btf;
   }
+  else
+    load_kernel_btfs(modules);
 
-  ssize_t ret = fread(data, 1, st.st_size, f);
-
-  if (ret != st.st_size)
+  if (btf_objects.empty())
   {
-    free(data);
-    fclose(f);
-    return nullptr;
+    if (bt_debug != DebugLevel::kNone)
+      LOG(ERROR) << "BTF: failed to find BTF data";
+    return;
   }
 
-  fclose(f);
-
-  *sizep = size;
-  return data;
-}
-
-static struct btf* btf_raw(char *file)
-{
-  unsigned char *data;
-  ssize_t size;
-  struct btf *btf;
-
-  data = get_data(file, &size);
-  if (!data)
-  {
-    LOG(ERROR) << "BTF: failed to read data from: " << file;
-    return nullptr;
-  }
+  vmlinux_btf_size = (__s32)type_cnt(vmlinux_btf);
 
-  btf = btf__new(data, (__u32) size);
-  free(data);
-  return btf;
+  state = OK;
 }
 
-static int libbpf_print(enum libbpf_print_level level, const char *msg, va_list ap)
+BTF::~BTF()
 {
-  fprintf(stderr, "BTF: (%d) ", level);
-  return vfprintf(stderr, msg, ap);
+  for (auto &btf_obj : btf_objects)
+    btf__free(btf_obj.btf);
```

Comment:
If you wanna get C++ fancy, you can put them in 
`std::unique_ptr<struct btf, void (*)(struct btf *)>` like this:
```
std::unique_ptr<struct btf, void (*)(struct btf *)> btf_obj{btf, &btf__free};
```

Reply:
While I'm usually a fan of this kind of solutions, it would just unnecessarily complicate the code in this case.